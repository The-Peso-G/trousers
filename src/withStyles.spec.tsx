import React from 'react';
import { render, cleanup } from '@testing-library/react';

import {
    css,
    withStyles,
    WithStylesProps,
    ThemeProvider,
    styleCollector,
} from './';

interface Props extends WithStylesProps {
    isRed?: boolean;
}

describe('withStyles', () => {
    afterEach(() => {
        document.getElementsByTagName('html')[0].innerHTML = '';

        cleanup();
    });

    it('attaches styles to the head', () => {
        const styles = styleCollector<Props, {}, {}>('foo').element`
            background-color: blue;
        `;

        const StyledComponent = withStyles(({ className }) => {
            return <span className={className}>Ahoy!</span>;
        }, styles);

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches styles with modifiers to the head', () => {
        const styles = styleCollector<Props, {}, {}>('foo').element`
            background-color: blue;
        `.modifier(props => !!props!.isRed)`
            background-color: red;
        `;

        const StyledComponent = withStyles(({ className }) => {
            return <span className={className}>Ahoy!</span>;
        }, styles);

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(<StyledComponent isRed />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches themed styles to the head', () => {
        interface Theme {
            backgroundColor: string;
            backgroundColorSecondary: string;
        }

        const theme: Theme = {
            backgroundColor: 'purple',
            backgroundColorSecondary: 'green',
        };

        const styles = styleCollector<Props, {}, Theme>('foo').element`
            background-color: ${theme => theme.backgroundColor};
        `.modifier(props => !!props!.isRed)`
            background-color: ${theme => theme.backgroundColor};
        `;

        const StyledComponent = withStyles(({ className }) => {
            return <span className={className}>Ahoy!</span>;
        }, styles);

        render(
            <ThemeProvider theme={theme}>
                <StyledComponent />
            </ThemeProvider>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(
            <ThemeProvider theme={theme}>
                <StyledComponent isRed />
            </ThemeProvider>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches single style to the head', () => {
        const StyledComponent = withStyles<Props, {}>(
            ({ className }) => <span className={className}>Ahoy!</span>,
            css`
                background-color: blue;
            `,
        );

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches themed single style to the head', () => {
        interface Theme {
            backgroundColor: string;
        }

        const theme = {
            backgroundColor: 'Atomic Tangerine',
        };

        const StyledComponent = withStyles<Props, Theme>(
            ({ className }) => <span className={className}>Ahoy!</span>,
            css<Theme>`
                background-color: ${theme => theme.backgroundColor};
            `,
        );

        render(
            <ThemeProvider theme={theme}>
                <StyledComponent />
            </ThemeProvider>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });
});
