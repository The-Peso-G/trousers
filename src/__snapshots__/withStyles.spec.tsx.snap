// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withStyles attaches single style to the head 1`] = `
<head>
  <style
    data-trousers=".css__293439549"
    type="text/css"
  >
    .css__293439549{background-color:blue;}

  </style>
</head>
`;

exports[`withStyles attaches styles to the head 1`] = `
<head>
  <style
    data-trousers=".foo__3889083325"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

  </style>
</head>
`;

exports[`withStyles attaches styles with modifiers to the head 1`] = `
<head>
  <style
    data-trousers=".foo__3889083325"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

  </style>
</head>
`;

exports[`withStyles attaches styles with modifiers to the head 2`] = `
<head>
  <style
    data-trousers=".foo__3889083325 .foo--491189580"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

    .foo--491189580{background-color:red;}

  </style>
</head>
`;

exports[`withStyles attaches themed single style to the head 1`] = `
<head>
  <style
    data-trousers=".css__36256131553516409327"
    type="text/css"
  >
    .css__36256131553516409327{background-color:Atomic Tangerine;}

  </style>
</head>
`;

exports[`withStyles attaches themed styles to the head 1`] = `
<head>
  <style
    data-trousers=".foo__38783974113106931827"
    type="text/css"
  >
    .foo__38783974113106931827{background-color:purple;}

  </style>
</head>
`;

exports[`withStyles attaches themed styles to the head 2`] = `
<head>
  <style
    data-trousers=".foo__38783974113106931827 .foo--38783974113106931827"
    type="text/css"
  >
    .foo__38783974113106931827{background-color:purple;}

    .foo--38783974113106931827{background-color:purple;}

  </style>
</head>
`;
